smcontroller/smcontroller.go

```diff
@@ -124,48 +141,37 @@ func New(
 		}
 	}
 
+	pb.RegisterSMServiceServer(controller.grpcServer, controller)
+
 	controller.logHandler = map[string]func(cloudprotocol.RequestLog) error{
 		cloudprotocol.SystemLog:  controller.getSystemLog,
 		cloudprotocol.ServiceLog: controller.getServiceLog,
 		cloudprotocol.CrashLog:   controller.getCrashLog,
 	}
 
-	var opts []grpc.ServerOption
+	if err := controller.startGRPCServer(); err != nil {
+		return nil, err
+	}
 
 	if !insecureConn {
-		certURL, keyURL, err := certProvider.GetCertificate(cfg.CertStorage, nil, "")
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
-		}
+		var ch <-chan *iamanager.CertInfo
 
-		tlsConfig, err := cryptcoxontext.GetClientMutualTLSConfig(certURL, keyURL)
-		if err != nil {
+		if ch, err = controller.certProvider.SubscribeCertChanged(cfg.CertStorage); err != nil {
 			return nil, aoserrors.Wrap(err)
 		}
 
-		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
-	} else {
-		log.Info("GRPC server starts in insecure mode")
+		go controller.processCertChange(ch)
 	}
 
-	controller.grpcServer = grpc.NewServer(opts...)
-
-	pb.RegisterSMServiceServer(controller.grpcServer, controller)
-
-	go func() {
-		if err := controller.startServer(cfg.SMController.CMServerURL); err != nil {
-			log.Errorf("Can't start SM controller server: %v", err)
-		}
-	}()
-
 	return controller, nil
 }
 
 // Close closes SM controller.
 func (controller *Controller) Close() error {
 	log.Debug("Close SM controller")
 
-	controller.stopServer()
+	controller.closeChannel <- struct{}{}
+	controller.stopGRPCServer()
```

Comment:
Should stopGRPCServer be guarded with mutex to do not conflict with reconnect?

Reply:
It uses GRPCServer.stop from grpchelpers that is already guarded